src/math.rs

```diff
@@ -67,3 +67,141 @@ impl Cube {
         )
     }
 }
+
+#[derive(Debug, Clone)]
+pub struct OrientedBeam {
+    // The members here are an implementation detail and differ from the
+    // minimal representation in the gRPC message to speed up operations.
+    // Rotation_inv and translation_inv together form the transform from world
+    // coordinates into "beam coordinates".
+    rotation_inv: Quaternion<f64>,
+    translation_inv: Vector3<f64>,
+    half_extent: Vector2<f64>,
+    corners: [Point3<f64>; 4],
+    separating_axes: [Vector3<f64>; 5],
+}
+
+impl OrientedBeam {
+    pub fn new(
+        rotation: Quaternion<f64>,
+        translation: Vector3<f64>,
+        half_extent: Vector2<f64>,
+    ) -> Self {
+        OrientedBeam {
+            rotation_inv: rotation.conjugate(),
+            translation_inv: -rotation.conjugate().rotate_vector(translation),
+            half_extent,
+            corners: OrientedBeam::precompute_corners(&rotation, &translation, &half_extent),
+            separating_axes: OrientedBeam::precompute_separating_axes(&rotation),
+        }
+    }
+    pub fn intersects(&self, aabb: &Aabb3<f64>) -> bool {
+        // SAT algorithm
+        // https://gamedev.stackexchange.com/questions/44500/how-many-and-which-axes-to-use-for-3d-obb-collision-with-sat
+        let mut separated = false;
+        for sep_axis in self.separating_axes.iter() {
+            // Project the cube and the beam onto that axis
+            let mut cube_min_proj = std::f64::MAX;
+            let mut cube_max_proj = std::f64::MIN;
+            for corner in aabb.to_corners().iter() {
+                let corner_proj = corner.dot(*sep_axis);
+                cube_min_proj = cube_min_proj.min(corner_proj);
+                cube_max_proj = cube_max_proj.max(corner_proj);
+            }
+            // Project four "vertices" of the beam onto that axis
+            let mut beam_min_proj = std::f64::MAX;
+            let mut beam_max_proj = std::f64::MIN;
+            for corner in self.corners.iter() {
+                let corner_proj = corner.dot(*sep_axis);
+                beam_min_proj = beam_min_proj.min(corner_proj);
+                beam_max_proj = beam_max_proj.max(corner_proj);
+            }
+            separated |= beam_min_proj > cube_max_proj || beam_max_proj < cube_min_proj;
```

Comment:
In case `separated` is `true`, we can break early here, right?

Reply:
Yes